ADR on modelling history redux

kevindew opened the pull request:
Trello: https://trello.com/c/Xt19LlP7/484-an-adr-for-editioning-versioning-documents

👉 [View rendered version](https://github.com/alphagov/content-publisher/blob/modelling-history-adr/docs/adr/0009-modelling-history.md) 👈

I've decided to close the previous PR as there was lots of discussion there that was now somewhat dated. I've updated various parts of this ADR to reflect changes that were made and I've removed the numbering aspect of it.

emmabeynon commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+- Consider how a revision of a document and aspects of it have state that may
+  change;
+  *because* the concept of revision could become convoluted if it stores
+  information that a user does not directly impact.
```
Could you give an example of such information?

1pretz1 replied:
+1 to the above comment as I don't really understand this point?

emmabeynon replied:
Ok got it.  Perhaps we could say something like "the concept of revision could become convoluted if it stores information that is not the content itself, such as internal state data"?

kevindew replied:
Yup that sounds good - I've now dropped this section as it generated a lot of comments/confusion and hopefully is superfluous to the overall document.

1pretz1 commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+  *because* we may need to re-present either to the Publishing API as Content
+  Publisher is the source of truth.
+- Consider how a revision of a document and aspects of it have state that may
+  change;
```
Don't really understand this point?

1pretz1 commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+This modelling now has a concept of:
+
+**Status**: Represents a particular singular state that an edition can
```
Maybe lose the word 'singular' as I think its clearer to just put `Represents a particular state that an edition can hold ...`

kevindew replied:
I went for the word "singular" to represent that it's meant to be just a single word that describes that edition as a whole. 

So if an edition is a "draft" it can't be "submitted for review", they're separate things. Whereas another aspect of state is `revision_synced` but that's not a singular thing as it can be applied to multiple statuses.

Though as I type this I don't think I'm doing too good a job of explaining here so it might be actually better to try work out how to rephrase it. I think what I'm trying to explain is that compared to other stateful aspects of an edition a status is special because it's 1) shown to the user 2) used for filtering and 3) can only be in one of them at a time.

kevindew replied:
I've had a go at rephrasing this section now.

1pretz1 commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+timeline to users visually and not for storing historical metadata.
+
+The TimelineEntry model should not store data which could not be
+derived from other aspects of a document. This is to maintain
```
`The TimelineEntry model should not store data which could not be derived from other aspects of a document` -  could you give an example of what data this is referring to?

kevindew replied:
We had a previous example of this where Removal, Retire and InternalNote were associated to a TimelineEntry rather than to a document directly.

This meant that when we wanted to show that a document was removed we had to query timeline entry to get the first removal. So in this instance TimelineEntry was being queried to find out what the removal was for an Edition

kevindew replied:
I've included that information in the doc now.

1pretz1 replied:
Nice!

1pretz1 commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+Some ideas to alleviate this are:
+
+- the use of [delegate][] with ActiveRecord to make it simpler to access data
+  without interacting with individual models;
```
A note on this - I agree with the use of delegate but can be quite confusing when it's nested through multiple models

kevindew replied:
Yup absolutely, they're definitely a trade off and with each level they get harder to trace to origin.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+### Could
+
+- Link TimelineEntry to changes rather than storing the information twice;
```
Not necessarily - you could change status for instance which isn't a revision change.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+- Enable a future feature to enable reverting to different points in time;
+  *because* this could enable users to make minor changes to a live document
+  and then recover a draft.
+- Separate related data to different tables to reduce data usage;
```
Sure, this'll be another area where I struggled to be succinct.

It's based on the idea of storing bits of revisions in different tables that are unlikely to change at the same time so that a revision itself isn't a huge amount of data.

I'll have a think about better ways to phrase it.

kevindew replied:
I've dropped this section now

kevindew commented:
Thanks @emmabeynon and @1pretz1. @emmabeynon I really appreciated your English Teacher hat 👍.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+  the Publishing API. We'd want to inform the user that a transition was in
+  progress and whether that failed.
+
+An approach to resolve this would be to provide the ability for an edition to
```
This is something that the design/product side are having to consider with regards to scheduling so the decision on that will be likely dictated from that.

My logic is that you're in the same status until the status change completes. E.g. if you're scheduled you're still there until publishing completes, if you're unpublishing you're published until that completes. These might not all hold though.

Pending kind of works unless the pending fails and then you end up in a state of published but not actually published. It could even be that there's a concept of a cover all "failed send help" status that would then allow a pending concept. 

Key outcome I'm hoping for with the scheduling work is that we're able to find a pattern that is suitable to scale to asynchronously allowing publishing/unpublishing/etc.

benthorner replied:
Thanks for explaining more :-). I think our logic is quite similar; either way we end up with a piece of state - whether it's a pending flag on a status, or a separate pending status, that we need to do something with. Since we have to do work either way in case of failure/success, it seems to me like the flag approach is simpler, but perhaps I'm missing some advantage?

kevindew replied:
No no, you're correct that a single status with a flag would be simpler.

It does feel a bit weird to me that you'd have something with a status of "published" that isn't actually published but if that fits into how product understands it then that'd be fine.

But yeah to me the key bit is what something is when a status transition fails, that approach there will probably dictate what option is applicable.

Btw I'm planning to drop this section from the ADR, but very happy to continue discussing our best options for this as this would be a really good problem to solve.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+but a new join between them is. A revision has a number of components which are
+described further in this document.
+
+This modelling now has a concept of:
```
I've managed to get rid of it 👍

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+The suggestion is that we enhance Content Publisher to support the concepts of
+Document, Edition and Revision.
+
+For a document with a live version and a draft this could enable:
```
Yup, good point. I've reproached this section.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+Content Publisher currently has a concept of a document and not editions or
+revisions. A consequence of this is that an edit to a document overwrites the
+previous state of the document. Since a document can exist in two forms, draft
+and live, we can only know about one of these at any time and not of any
```
I've rewrote this section.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+revisions. A consequence of this is that an edit to a document overwrites the
+previous state of the document. Since a document can exist in two forms, draft
+and live, we can only know about one of these at any time and not of any
+previous versions.
```
Yup good point, I've added that into the re-approach of context.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+### Must
+
+- Store full information on current and live versions of a document;
```
I've dropped this section now. I think I make use of your snapshot suggestion later though.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+- Store full information on current and live versions of a document;
+  *because* we may need to re-present either to the Publishing API as Content
+  Publisher is the source of truth.
+- Consider how a revision of a document and aspects of it have state that may
```
Good points, I've dropped this bit now. It was much more relevant as part of scoping that the actual decision.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+  *because* this will allow us the means to present what changed between
+  published editions - which is needed for a variety reasons, including, in
+  some cases, legal.
+- Record every change a publisher makes and allow for that information to be
```
I think I was badly referencing how hard it would have been to access that data through papertrail. That's gone now.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+  some cases, legal.
+- Record every change a publisher makes and allow for that information to be
+  accessible;
+  *because* this maintains options for TimelineEntry and provides rich
```
Dropped this section now.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+  accessible;
+  *because* this maintains options for TimelineEntry and provides rich
+  debugging information.
+- Avoid relying on the TimelineEntry model as the main answer to any
```
I've dropped this section now, hopefully I've captured the essence elsewhere.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+### Could
+
+- Link TimelineEntry to changes rather than storing the information twice;
+  *because* this would reduce concerns and opportunities for error.
```
Dropped section.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+live edition which is currently on GOV.UK. The live and current edition can be
+the same. Document serves as a mutable entity that tracks data across all
+editions (such as first publishing date) and it is expected to be a joining
+point for document related data that is not associated with a particular
```
Thanks - got it.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+**Edition**: A numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a revision
+and a status. It is mutable and is expected as a place where any edition level
```
Thanks - done so.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+common way to interact with append-only database models. This is done for
+simplicity, performance and for consistency with Rails idioms. To achieve this
+we maintain a number of mutable models (such as Document or Edition) that are
+updated to the most recent foreign keys which makes them not suitable vessels
```
I've rephrased this bit now.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+The intention of breaking this up is to be conservative with the amount of data
+duplicated for both storage and maintenance reasons. Thus when a user edits
+the title of an edition a new ContentRevision object is created and the other
```
Thanks 👍

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+The data of an Image::Revision is stored between an Image::FileRevision and an
+Image::MetadataRevision. These two objects differ by the fact that any change to
+Image::FileRevision requires changes to the resultant files (such as crop
```
I've changed it to Asset Manager files. I felt "asset" was also rather ambiguous.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+many association with Image::Asset, each of these represent resulting files that
+are uploaded to Asset Manager. For a single image uploaded we'd likely upload
+multiple sizes to Asset Manager - these different files are named variants. An
+Image::Asset object stores whether the asset is on Asset Manager, what state it
```
I've rephrased this.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+A timeline entry represents an event that should be shown to a user as part
+of a documents history. This stores the event which occurred and associations
+to relevant models. The sole purpose of this is for representing a
```
Good point, I've amended this.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+Further details are then explained in the following sections.
+
+### Approach to mutability/immutability
```
Yeah I can understand you got lost. I didn't really like the "Decision point: " prefix but I've hopefully improved this by reducing the number of headers and overall scope of the document.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,314 @@
+
+Using terminology as defined in [ADR-3](0003-initial-domain-modelling.md):
+
+- **Document**: A piece of content in a particular locale that can have many
```
Sure, I've removed it now.

kevindew commented:
@emmabeynon @1pretz1 @benthorner I've done a fair re-working of this (which probably means there's a fresh batch of English errors 😬) let me know what you think.

emmabeynon commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+And this prevents a number of intended features for Content Publisher:
+
+- comparing different editions of a document;
+- republishing live content if there is any problems - which is a frequent
```
Suggested rephrasing:
"republishing live content if there are any problems (currently a common support task for Whitehall publisher)"

emmabeynon commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+particular locale. It has many editions and at any time it will have a current
+edition - shown on Content Publisher index - and potentially a live edition
+which is currently on GOV.UK. The live and current edition can be
+the same. Each iteration of a documents content is represented as a revision,
```
"documents" => "document's"

emmabeynon commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+No data related to topics (otherwise known as GOV.UK taxonomy) is intended to
+be stored in Content Publisher at the current point in time. This is due to
+topics being accessed and edited by directly interacting with the Publishing
+API. This is because other applications, notably Content Tagger, can also edit
```
Delete "This because"

emmabeynon commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+be stored in Content Publisher at the current point in time. This is due to
+topics being accessed and edited by directly interacting with the Publishing
+API. This is because other applications, notably Content Tagger, can also edit
+these which has the consequence that the Publishing API is the source of truth
```
"these" => "topics"

emmabeynon commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+a document and an images table which stores the most recent version of
+a document's images. History of these changes is stored using [Papertrail][],
+which is not [intended as a permanent store][papertrail-pr]. The presentation
+of a documents history is done via a TimelineEntry model which stores that a
```
"documents" => "document's"

kevindew commented:
Thanks @emmabeynon - I've made those changes.

1pretz1 commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+This has led to a number of pain points:
+
+- users cannot discard a draft of a document that has between previously
```
I think you mean `been` here instead of `between` or `has previously been published` ?

kevindew replied:
Oops! thanks 👍

1pretz1 commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+Content Publisher supports a user uploading image files and referencing them
+in a revision of a document. They have metadata and editable properties that a
+user can change, of which a history stored. A single image file uploaded
```
I don't think this reads quite right maybe `which a history stored` -> `which a history is stored`

kevindew replied:
Oops again - thanks. I don't know why I'm so bad at English 😞

1pretz1 commented:
@kevindew Good stuff, this is much clearer / easier to understand! I've just left a few comments other then that it LGTM 👍

kevindew commented:
Thanks  @1pretz1 I've made those changes.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+### Approach to mutability/immutability
+
+A number of the models in Content Publisher are defined as immutable, most
+significantly Revision and its associated types. These models should be
```
I think we should be more specific about 'type' here, or perhaps link to the section that explains more. Fortunately, 'type' isn't currently a concept we're using in the app - long may this last!

kevindew replied:
swapped "type" for "models"

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+Accessing immutable models is intended to always be done by a foreign key and
+not by the usage of `SELECT MAX` style queries. This is done for
+simplicity, performance and for consistency with Rails idioms. To achieve this
+there are a some mutable models, eg Edition, that reference immutable models,
```
eg -> like (or e.g.)

kevindew replied:
eg is GOV.UK style

I've rephrased this section and it no longer has the eg

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+them requires creating a new record. This allows us to store a full history
+by only appending to the database.
+
+Accessing immutable models is intended to always be done by a foreign key and
```
Could we think about rephrasing this paragraph, so it explains the problem before the solution? The wording is also a bit strange, as it suggests we 'invented' the Edition model to solve this particular problem, whereas I think it's intended for illustration purposes.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+there are a some mutable models, eg Edition, that reference immutable models,
+eg Revision. When the content for an edition is revised the Edition model
+references the new Revision. Mutable models should not be used to store fields
+where there is a need to know how the value of that field changed over time.
```
Could we simply say "Mutable models should not store data that can have a history"? Earlier, we talked about whether something is "changeable by the user directly" - is that appropriate to re-iterate here?

kevindew replied:
I've added a paragraph covering this.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+where there is a need to know how the value of that field changed over time.
+
+The choice of this immutability strategy is to store both present and
+historical concerns in the same data stores, thus ensuring history remains a
```
I read 'data store' as the choice of database. How about "is to model current and historical data in the same way"?

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+### Breakdown of Revision
+
+Revision is a model that stores little data itself and joins to other models
```
Would be good to emphasise the immutability here, as this is key to the following argument...

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+![Revision breakdown](000x/revision-diagram.png)
+
+The intention of breaking this up is to be conservative with the amount of data
+duplicated for both storage and maintenance reasons. Thus when a user edits
```
"for both storage and maintenance reasons" -> "when the content of the revision changes".

benthorner replied:
"Thus" -> "For example" (there's no particular implication here)

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+The intention of breaking this up is to be conservative with the amount of data
+duplicated for both storage and maintenance reasons. Thus when a user edits
+the title of an edition a new ContentRevision object is created and the existing
+objects (TagsRevision, MetadataRevision and ImageRevisions) are associated with
```
How about just "and the existing TagsRevision, ..."? I wonder whether it's worth casually introducing the term 'sub-revision', as 'object' seems overly generic for a model.

kevindew replied:
I've gone with naming the individual models.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+Content Publisher supports a user uploading image files and referencing them
+in a revision of a document. They have metadata and editable properties that a
+user can change, of which a history is stored. A single image file uploaded
```
Could we try "gets turned into multiple assets (byte streams)" instead? I think it's OK to be a bit more specific here, as all the concepts are quite meta ;-).

kevindew replied:
Asset is more ambiguous term than file and would do with qualifying. I understand you've tried to do that with the "byte streams" suggestion but I think that ends up opening a new question of "what is the relevance of a byte stream"?

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+![Image Revision breakdown](000x/image-revision-diagram.png)
+
+The Image model itself is used for continuation between image revisions. It is
```
Although this is true, it might be a good idea to give an example of why we need to know this.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+same Image association.
+
+The data of an Image::Revision is stored between an Image::FileRevision and an
+Image::MetadataRevision. These two objects differ by the fact that any change to
```
'two objects' -> 'two models'

benthorner replied:
Could we rephrase to say the less important thing first? Something like 'Both have the same immutable behaviour as the sub-revisions of a Revision, but any change to...'

benthorner replied:
Could we use 'assets' instead of 'Asset Manager files', as the latter doesn't seem to be defined - it's at least unclear whether it's a concept in our app or external (or both ;-).

kevindew replied:
I've changed this around a bit.

As per other comment, you have a point but I think "asset" requires more qualifying and comparing the terms in context we mention "files that are uploaded to Asset Manager" in the previous sentence so this terminology follows.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+Image::FileRevision is associated with an ActiveStorage::Blob object that is
+responsible for managing the storage of the source file. It also has a one to
+many association with Image::Asset, each of these represent resulting files that
+are uploaded to Asset Manager. The Image::Asset model stores the URL to the
```
Probably good to reiterate the size thing here somewhere, to clarify why it's a one-to-many.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+The TimelineEntry model represents an event that should be shown to a user as
+part of a visual timeline of a document's history. It stores the event which
+occurred and associations to relevant models. The sole purpose of this model
```
Perhaps add 'polymorphic' here to indicate the association is very flexible.

benthorner replied:
The reasoning is getting a bit lost here in trying to explain how things used to be. I think this change might help: 

'The sole purpose...' -> 'At the time of writing it wasn't yet determined what the timeline would show, and therefore it wasn't clear exactly how best to model them. In order to support future iteration on this model, it should not be used outside of the context of displaying a timeline'.

We could then modify the following paragraph to explain one of the consequences of this is/would-be to re-think the associations for Removals and Withdrawals.

kevindew replied:
I've changed this around.

I've amended your suggestion though as that implies that we don't intend to use it for other reasons because we don't know the future iteration. Whereas that is only one of the reasons, the other being separation of concerns.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+At the time of writing it wasn't yet determined what the
+timeline would show, and therefore it wasn't clear exactly how
+best to model them. Because of this TimelineEntry will be modelled in a
```
'model them' -> 'model an entry for it'

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+### Topics
+
+No data related to topics (otherwise known as GOV.UK taxonomy) is intended to
+be stored in Content Publisher at the current point in time. This is due to
```
'the' -> 'this'

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+This inconsistency makes it difficult to store the history of topics in a
+reliable way. Thus, until needs are determined to store this data, Content
+Publisher will continue to query the Publishing API rather than storing this
+data.
```
...and history for topics is not supported. Otherwise, with the previous sentence, it sounds like we're currently querying for history.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+the next revision. An ImageRevision is modelled in a similar way to a Revision
+and this is explained further in [Image modelling](#image-modelling).
+
+It is intended that [delegation][delegate] be used when interfacing with a
```
Should we also mention that we intend to delegate update operations, for the same reason?

kevindew replied:
Oh sorry - I thought this was in reference to the models section in consequences which I updated.

Looking at this I don't think it needs changing as it only references interfacing which can refer to both reading or writing.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+Accepted
+
+## Consequences
```
I wonder if this section should discuss ideas for migration to the new modelling, as this was definitely a design decision. It does feel somewhat pointless now that it's actually implemented, but it's curious to have so few consequences for such a big proposal; the missing piece seems to be the implementation/migration strategy.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+information. There is a documents table which stores the most recent content of
+a document and an images table which stores the most recent version of
+a document's images. History of these changes is stored using [Papertrail][],
+which is not [intended as a permanent store][papertrail-pr]. The presentation
```
I'm not sure what you're suggesting?

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+- users cannot discard a draft of a document that has between previously
+  published because there isn't a data representation it can be reverted to;
+- Content Publiser can't show an accurate link or status for the live edition
```
You missed by big "Publiser" typo 😁 

I've amended this with different wording than what you suggested.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+  of a document;
+- users cannot edit or remove images on a document once the first
+  edition is published;
+- the TimelineEntry model stores aspects of a documents state, resulting in it
```
Thanks - my English is shocking 😢 

I've added a quick "which.." to this to clarify. It's a bit of a balancing act trying to make the list succinct when there's lot of detail that could be contained.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+- comparing different editions of a document;
+- republishing live content if there are any problems (currently a common
+  support task for Whitehall publisher);
+- showing users what changes a user made in a particular edit.
```
Only in the sense that to us they're both different ways of comparing a revision.

In most of our publishing apps these are quite distinct features though where edition to edition comparison is common and revision to revision change is done manually (for example mainstream publisher, where a publisher writes out the diff).

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+This ADR proposes changes to the domain model to resolve the aforementioned
+pain points and provide a means to support the future intended features. It
+defines a number of [core concepts](#core-concepts) that underpin the
```
It's not intended to be a link to the image it's intended to be a link to the section. This section is supposed to introduce and summarise the ADR. I realise linking to the first thing isn't particularly useful so stripped that out.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+[whether models are mutable](#approach-to-mutabilityimmutability) once
+persisted. Immutability is favoured as a means to preserve a history since
+each change requires storing a new record, implicitly creating a history.
+Mutability is used as a means to simplify interacting with models and preserve
```
Initially I made your edit like you suggested, however it bugged me as this was the one use of "we" in the document. 

In the end I've only rephrased this a bit, hopefully you deem it an improvement.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+Publisher will continue to query the Publishing API rather than storing this
+data.
+
+### Overview
```
Yeah summary has the same problem has overview in that it's frequently used at the introduction rather than conclusion of a document. I purposely opted out of summary since that carried connotations of brief and simplified since this is probably the most complicated part.

For lack of a better term I've gone with "Collated Diagram" which may well be the worst of all worlds. If you have any other suggestions that'd be great. I'm looking for something that reflects that it's the combined picture which I was what I hoped "Overview" portrayed.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+Mutability is used as a means to simplify interacting with models and preserve
+Rails idioms.
+
+This ADR then considers the impact of history for [timeline](#timeline) and
```
It's meant to act as a way of summarises all that's covered in the decision. I think without it this introduction loses flow.

Maybe there is just too much going on in the mutability paragraph before, that might be what makes it a less smooth read?

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+edition - shown on Content Publisher index - and potentially a live edition
+which is currently on GOV.UK. The live and current edition can be
+the same. Each iteration of a document's content is represented as a revision,
+thus a document has many revisions. Document is a mutable entity that tracks
```
I've rephrased the use of track. I think it's useful to say that it is intended to store data common to the Document.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+and a status. It is mutable and is a place where any edition level
+database constraints can be placed, such as only one live edition can exist for
+a document. It is possible that two editions of the same document share the
+same revision, this is in the case that they have identical content.
```
I struggled how to resolve this concern of yours for a while.

I don't think imply is the correct term as it's explicit rather than implicit that the content is the same. I've tried to capture this in a sentence.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+
+**Edition**: A numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a revision
+and a status. It is mutable and is a place where any edition level
```
I've added a clause in to say why.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+particular point in time. It has a number to indicate which revision of the
+document it is and stores who created it. Any request by a user that changes
+content should result in a single new revision, and only data that is
+changeable by a user directly should be stored as part of a revision. This is
```
Fair point.

I was referring to Revisions should not be used to hold data that may change independently of a users request. For example, were we to store Asset Manager state inside a revision and created new revisions each time we changed data in Asset Manager. That would be a violation of a Revision to me as it would mean that our history is confused by events that a user created and events our system created. This would prevent us from ever presenting revision history to users or revision numbers.

I've tried to rephrase this based around what we do forbid rather than what we should include.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+a new Status model is created and the user who created it stored. An edition
+can only have one status at any one time. It is intended to associate with
+specialist data for a state such as a Removal or Withdrawal, which allows the
+history of these to no longer be the responsibility of TimelineEntry. Initially
```
I don't think that's necessary, the content here represents our current understanding and changing that wouldn't be an accurate reflection on the intentions of the model.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,250 @@
+"submitted for review". This model is coupled to the concept of status that is
+shown and changed by a user. Each time a user changes the status of an edition
+a new Status model is created and the user who created it stored. An edition
+can only have one status at any one time. It is intended to associate with
```
There isn't a section that explains more, I've added more information here though to try make it less mysterious.

kevindew commented:
With much concern that this will only lengthen this rather protracted review cycle I've added a couple of new consequences section in.

@emmabeynon I've added a new section to the consequences about the wrapping of actions inside transactions. Do let me know if that reads ok to you and if it helps explain the "why"

@benthorner You're right about the missing consequence, it's a part that is rather weird in hindsight. I've added an extra section in that addresses the tech debt part as that seems to be the part that is currently relevant.

benthorner commented:
Nearly there, I think :-). The new consequences section looks good - most of my comments are follow-ups on the sections before.

emmabeynon commented:
> @emmabeynon I've added a new section to the consequences about the wrapping of actions inside transactions. Do let me know if that reads ok to you and if it helps explain the "why"

Reads well to me 👍

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,343 @@
+**Edition**: A numbered version of a document that has been, or is
+expected to be, published on GOV.UK. It is associated with a revision
+and a status. It is mutable so that it can be a consistent object that
+joins to immutable data. It is a place where any edition-level
```
I don't think that's the reason, as we could do the same through a series of immutable models. To me it is just so that it is a singular thing and can manage constraints.

kevindew commented on docs/adr/000x-modelling-history.md:
```diff
@@ -0,0 +1,343 @@
+and a status. It is mutable so that it can be a consistent object that
+joins to immutable data. It is a place where any edition-level
+database constraints can be placed, such as only one live edition can exist for
+a document. It is supported that two editions of the same document share the
```
I don't think the scenario you've picked there is a particularly pertinent one as withdrawn doesn't make a difference. I've stayed away from referring to last revision of live edition being equal to first revision of next draft because of the change note stuff, so instead I've referenced the potential future revert aspect.

kevindew commented:
Thanks @emmabeynon 

@benthorner I've addressed most of your comments in https://github.com/alphagov/content-publisher/pull/729/commits/a3bab976c006df3b639b33f17d858586b7b831b1 and replied to ones I've not addressed.

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -93,27 +93,31 @@ this document.
-accessing an edition by a consistent primary key and the access of it's
-revision by a foreign key on the edition.
+of `SELECT MAX` style queries. This maintains the ability to use the regular
+approach to ActiveRecord associations and the means to require an associations
```
"associations" -> "association's"

benthorner commented on docs/adr/000x-modelling-history.md:
```diff
@@ -172,10 +176,10 @@ Manager files (such as crop dimensions), whereas Image::MetadataRevision stores
-that are uploaded to Asset Manager for the various image sizes. The
+Each Image::FileRevision is associated with an ActiveStorage::Blob object that
+is responsible for managing the storage of the source file. It also has a one
+to many association with Image::Asset, each Image::Asset represents resultant
```
New sentence after ','

kevindew commented:
Finally done :tada: just shy of the 200 comments
